frameworks/non-keyed/kobold/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+    "odd",
+    "unsightly",
+    "adorable",
+    "important",
+    "inexpensive",
+    "cheap",
+    "expensive",
+    "fancy",
+];
+
+const COLOURS_LEN: usize = 11;
+static COLOURS: &[&str; COLOURS_LEN] = &[
+    "red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black",
+    "orange",
+];
+
+const NOUNS_LEN: usize = 13;
+const NOUNS: [&str; NOUNS_LEN] = [
+    "table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger",
+    "pizza", "mouse", "keyboard",
+];
+
+fn random(max: usize) -> usize {
+    (Math::random() * 1000.0) as usize % max
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct RowData {
+    id: usize,
+    label: String,
+}
+
+struct State {
+    rows: Vec<RowData>,
+    selected: Option<usize>,
+    last: usize,
+}
+
+impl State {
+    fn new() -> Self {
+        State {
+            rows: Vec::new(),
+            selected: None,
+            last: 0,
+        }
+    }
+    fn append_rows(&mut self, count: usize) {
+        self.rows.reserve_exact(count);
+        let d = self.last + 1;
+        for i in 0..count {
+            let adjective = ADJECTIVES[random(ADJECTIVES_LEN)];
+            let colour = COLOURS[random(COLOURS_LEN)];
+            let noun = NOUNS[random(NOUNS_LEN)];
+            let mut label = String::with_capacity(adjective.len() + colour.len() + noun.len() + 2);
+            label.push_str(adjective);
+            label.push(' ');
+            label.push_str(colour);
+            label.push(' ');
+            label.push_str(noun);
+            self.rows.push(RowData {
+                id: i + d,
+                label,
+            });
+        }
+        self.last += count;
+    }
+
+    fn remove(&mut self, id: usize) {
+        self.rows.remove(id);
+    }
+
+    fn run(&mut self) {
+        self.clear();
+        self.append_rows(1_000)
+    }
+    fn run_lots(&mut self) {
+        self.clear();
+        self.append_rows(10_000)
+    }
+    fn add(&mut self) {
+        self.append_rows(1_000)
+    }
+    fn update(&mut self) {
+        self.rows.iter_mut().step_by(10).for_each(|row| {
+            row.label.reserve_exact(4);
+            row.label.push_str(" !!!")
+        });
+    }
+    fn clear(&mut self) {
+        self.rows = Vec::new();
+        self.selected = None
+    }
+    fn swap(&mut self) {
+        if self.rows.len() > 998 {
+            self.rows.swap(1, 998);
+        }
+    }
+    fn select(&mut self, id: usize) {
+        self.selected = Some(id)
+    }
+}
+
+#[component]
+fn Row(num: usize, row: RowData, state: &Hook<State>) -> impl View {
+    let is_in_danger = class!("danger" if state.selected == Some(row.id));
+
+    bind! { state:
+        let remove = move |_| state.remove(num);
+        let select = move |_| state.select(row.id);
+    }
+
+    view! {
+        <tr.{is_in_danger}>
+            <td class="col-md-1">{row.id.to_string()}</td>
+            <td class="col-md-4"><a onclick={select}>{row.label}</a></td>
```

Comment:
```suggestion
            <td class="col-md-4"><a onclick={select}>{ref row.label}</a></td>
```

...which makes diffing by pointer here safe.

Reply:
Adding this ref makes the update button stop working correctly. Sometimes it increments the string sometimes it doesn't.

Reply:
FYI: I haven't checked it, but I think that's because the allocator is sometimes able to grow the allocation in place so the pointer doesn't change.

Could force a fresh allocation with some such:

```rust
// This creates a new string so that we temporarily have two, ensuring
// the new string has a different pointer.
row.label = String::from_iter([&row.label, " !!!"]);
```

Though that's super clumsy and I wouldn't expect user code to be written like this. Let's keep things as they are for now.